src/proper_statem.erl

```diff
@@ -462,6 +476,191 @@ move_shrinker(Seq, Par, I) ->
+
+-spec targeted_commands(mod_name(), symbolic_state()) -> proper_types:type().
+targeted_commands(Mod, InitialState) ->
+  proper_types:add_prop(is_user_nf_stateful, true,
+                        ?USERNF(targeted_commands_gen(Mod, InitialState),
+                                next_commands_gen(Mod, InitialState))).
+
+%% @private
+-spec targeted_commands_gen(mod_name()) -> proper_types:type().
+targeted_commands_gen(Mod) ->
+  ?LET(Cmds,
+       ?LAZY(commands(Mod)),
+       {finalize_weights(Mod, Cmds, maps:new()), Cmds}).
+
+%% @private
+-spec targeted_commands_gen(mod_name(), symbolic_state()) ->
+        proper_types:type().
+targeted_commands_gen(Mod, InitialState) ->
+  ?LET(Cmds,
+       ?LAZY(commands(Mod, InitialState)),
+       {finalize_weights(Mod, Cmds, maps:new()), Cmds}).
+
+%% @private
+-spec next_commands_gen(mod_name()) -> fun((_, _) -> proper_types:type()).
+next_commands_gen(Mod) ->
+  fun ({Weights, _Cmds}, {_D, _T}) ->
+      NewWeights = next_weights(Weights),
+      CmdsGen = next_gen(Mod, NewWeights),
+      ?SHRINK(
+         ?LET(Cmds, ?LAZY(CmdsGen),
+              {finalize_weights(Mod, Cmds, NewWeights), Cmds}),
+         [CmdsGen])
+  end.
+
+%% @private
+-spec next_commands_gen(mod_name(), symbolic_state()) ->
+        fun((_, _) -> proper_types:type()).
+next_commands_gen(Mod, InitialState) ->
+  fun ({Weights, _Cmds}, {_D, _T}) ->
+      NewWeights = next_weights(Weights),
+      CmdsGen = next_gen(Mod, InitialState, NewWeights),
+      ?SHRINK(
+         ?LET(Cmds, ?LAZY(CmdsGen),
+              {finalize_weights(Mod, Cmds, NewWeights), Cmds}),
+         [CmdsGen])
+  end.
+
+%% Produce the new random weights by adding or removing an integer
+%% value to the existing weights.
+
+%% @private
+-spec next_weights(weights()) -> weights().
+next_weights(Weights) ->
+  Min = -?RESIZE_FACTOR,
+  Max = ?RESIZE_FACTOR,
+  maps:map(fun (_Key, W) ->
+               Random = rand:uniform(1 + (Max - Min)) - (1 - Min),
+               NW = W + Random,
+               case NW >= 1 of
+                 true -> NW;
+                 false -> 1
+               end
+           end, Weights).
+
+%% Used to return the generator the commands of the targeted
+%% generation.
+
+%% @private
+-spec next_gen(mod_name(), weights()) -> proper_types:type().
+next_gen(Mod, Weights) ->
+  ?LET(InitialState, ?LAZY(Mod:initial_state()),
+       ?SUCHTHAT(
+          Cmds,
+          ?LET(List,
+               ?SIZED(Size,
+                      proper_types:noshrink(
+                        next_gen(Size * ?RESIZE_FACTOR, Mod,
+                                 InitialState, 1, Weights))),
+               proper_types:shrink_list(List)),
+          is_valid(Mod, InitialState, Cmds, []))).
+
+%% Used to return the generator the commands of the targeted
+%% generation with an initial state provided by the user.
+
+%% @private
+-spec next_gen(mod_name(), symbolic_state(), weights()) -> proper_types:type().
```

Comment:
Comment about `next_gen/2` and `next_gen/3`:
The convention/tradition in Erlang is that when there exist such functions the _additional_ arguments appear last in the function with arity+1.  Is this an issue here?